polygon/sync/header_downloader.go

```diff
@@ -0,0 +1,214 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/ledgerwatch/log/v3"
+
+	"github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/erigon/core/types"
+	"github.com/ledgerwatch/erigon/polygon/sync/peerinfo"
+)
+
+const headerDownloaderLogPrefix = "HeaderDownloader"
+
+func NewHeaderDownloader(logger log.Logger, sentry Sentry, db DB, heimdall Heimdall, verify HeaderVerifier) *HeaderDownloader {
+	statePointHeadersMemo, err := lru.New[common.Hash, []*types.Header](sentry.MaxPeers())
+	if err != nil {
+		panic(err)
+	}
+
+	return &HeaderDownloader{
+		logger:                logger,
+		sentry:                sentry,
+		db:                    db,
+		heimdall:              heimdall,
+		verify:                verify,
+		statePointHeadersMemo: statePointHeadersMemo,
+	}
+}
+
+type HeaderDownloader struct {
+	logger                log.Logger
+	sentry                Sentry
+	db                    DB
+	heimdall              Heimdall
+	verify                HeaderVerifier
+	statePointHeadersMemo *lru.Cache[common.Hash, []*types.Header] // statePoint.rootHash->[headers part of state point]
+}
+
+func (hd *HeaderDownloader) DownloadUsingCheckpoints(ctx context.Context, start uint64) error {
+	checkpoints, err := hd.heimdall.FetchCheckpoints(ctx, start)
+	if err != nil {
+		return err
+	}
+
+	err = hd.downloadUsingStatePoints(ctx, statePointsFromCheckpoints(checkpoints))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (hd *HeaderDownloader) DownloadUsingMilestones(ctx context.Context, start uint64) error {
+	milestones, err := hd.heimdall.FetchMilestones(ctx, start)
+	if err != nil {
+		return err
+	}
+
+	err = hd.downloadUsingStatePoints(ctx, statePointsFromMilestones(milestones))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (hd *HeaderDownloader) downloadUsingStatePoints(ctx context.Context, statePoints statePoints) error {
+	for len(statePoints) > 0 {
+		allPeers := hd.sentry.PeersWithBlockNumInfo()
+		if len(allPeers) == 0 {
+			hd.logger.Warn(fmt.Sprintf("[%s] zero peers, will try again", headerDownloaderLogPrefix))
+			continue
+		}
+
+		sort.Sort(allPeers) // sort by block num in asc order
+		peers := hd.choosePeers(allPeers, statePoints)
+		if len(peers) == 0 {
+			hd.logger.Warn(
+				fmt.Sprintf("[%s] can't use any peers to sync, will try again", headerDownloaderLogPrefix),
```

Comment:
@battlmonstr if this ever happens, ie we have all 100 max peers but all are not synced enough (highly unlikely I suppose but possible) we can add a sentry function to temporarily disconnect them for some time interval so we can connect to others that may be synced further ahead - what do you think? would a disconnect be the same as penalize in this case or penalized means we will never connect to the penalized peer again in the future?

Reply:
work for a future PR regardless